### PR TITLE
[Havoc] Initial Dash of Chaos Implementation

### PR DESCRIPTION
**Pull request by Remnxnt**
First attempt at dipping my toes in please let me know if you see anything out of line, thanks.

### CODE REVIEW

**Comment by vituscze** [comment]
Have you checked that this actually works? A lot of the conditions seem nonfunctional (you're checking whether talent's spell data pointer isn't `nullptr` rather than checking if the talent is enabled, for example).

**Comment by Remnxnt** [comment]
It does appear to be functional. Though there are some things that I couldn't quite figure out, such as the dash_of_chaos action is executing twice for every hit for example. This doesn't have an effect on the DPS contribution of the ability but does impact the DPE statistics. As for talent conditions you'd have to point out which one's specifically would be non-functional.

File: engine/class_modules/sc_demon_hunter.cpp [hunk 1]
```diff
@@ -5198,6 +5214,13 @@ struct fel_rush_t : public demon_hunter_attack_t
     // Fel Rush and VR shared a 1 second GCD when one or the other is triggered
     if ( p()->cooldown.movement_shared->down() )
       return false;
+   
+    // Not usable while Dash of Chaos buff is present and is locked out for 1 second upon buff expiration
+    if ( p()->talent.havoc.dash_of_chaos && p()->buff.dash_of_chaos && p()->buff.dash_of_chaos->check())
```
Comment (ToppleTheNun) [inline engine/class_modules/sc_demon_hunter.cpp]: `p()->talent.havoc.dash_of_chaos` needs to be `p()->talent.havoc.dash_of_chaos->ok()`

File: engine/class_modules/sc_demon_hunter.cpp [hunk 2]
```diff
@@ -5164,6 +5169,17 @@ struct fel_rush_t : public demon_hunter_attack_t
 
     p()->buff.unbound_chaos->expire();
 
+    if (p()->talent.havoc.dash_of_chaos->ok())
+    {
+      p()->buff.dash_of_chaos->trigger();
+    }
+
+    // Dash of chaos incurs 1 second lockout on Felrush
+    if ( p()->talent.havoc.dash_of_chaos && !p()->buff.dash_of_chaos->up() )
```
Comment (ToppleTheNun) [inline engine/class_modules/sc_demon_hunter.cpp]: `p()->talent.havoc.dash_of_chaos` needs to be `p()->talent.havoc.dash_of_chaos->ok()`

**Review by ToppleTheNun — COMMENTED** [review]
These are what were meant by checking if the talent is selected rather than `nullptr`.

**Comment by Remnxnt** [comment]
I did not intend to perform the second commit

File: engine/class_modules/sc_demon_hunter.cpp (hunk 2 above)
Comment (Remnxnt) [reply engine/class_modules/sc_demon_hunter.cpp]: this line has been removed in most recent commit

File: engine/class_modules/sc_demon_hunter.cpp (hunk 1 above)
Comment (Remnxnt) [reply engine/class_modules/sc_demon_hunter.cpp]: This has been fixed in most recent commit

**Comment by Remnxnt** [comment]
TODO: 

Add a method to cast dash_of_chaos while debuff_out_of_range is present, I suspect this is causing dash of chaos to be executed slower (0.7-0.9s) than is possible in game (0.5s)

Resolve issue of execute statistics (exactly double the amount of real casts)

**Pull request closed without merging**